<center>
    <br>
        <br>
            <br>
                <h5 style = "color:blue">====================================================================</h5>
                <h2 style = "color:green">NIPS Revise</h2>
                <h3 style = "color:green">( Random Forest Hyperparameter Tuning )</h3>
                <h5 style = "color:blue">====================================================================</h5>
            <br>
        <br>
    <br>
</center>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
!pip install tpot

     |████████████████████████████████| 87 kB 2.1 MB/s 
     |████████████████████████████████| 166.7 MB 10 kB/s 
     |████████████████████████████████| 160 kB 51.7 MB/s 
  Created wheel for stopit: filename=stopit-1.1.2-py3-none-any.whl size=11952 sha256=04fd232f7590a306b4dd2e3f9875d6700f44adc61931df5894a5667e09215c14
  Stored in directory: /root/.cache/pip/wheels/e2/d2/79/eaf81edb391e27c87f51b8ef901ecc85a5363dc96b8b8d71e3
Successfully built stopit
  Attempting uninstall: xgboost
    Found existing installation: xgboost 0.90
    Uninstalling xgboost-0.90:
      Successfully uninstalled xgboost-0.90


# Step 1: Import Libraries

In [3]:
# Import Libraries

import time
import xlwt
import warnings
import numpy as np
import pandas as pd
import seaborn as sns
from numpy import mean
from numpy import sqrt
from numpy import where
from numpy import argmax
from xlwt import Workbook
from sklearn.svm import SVC
from keras.layers import Dense
from sklearn.metrics import auc
from tpot import TPOTClassifier
from collections import Counter
warnings.simplefilter("ignore")
import matplotlib.pyplot as plt
plt.style.use('fivethirtyeight')
from sklearn import preprocessing
from prettytable import PrettyTable 
from sklearn import model_selection
from keras.models import Sequential
from sklearn.metrics import roc_curve
from imblearn.pipeline import Pipeline
from astropy.table import Table, Column
from imblearn.over_sampling import SMOTE
from sklearn.metrics import roc_auc_score
from sklearn.metrics import accuracy_score
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import cohen_kappa_score
from sklearn.ensemble import VotingClassifier
from sklearn.preprocessing import MinMaxScaler
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from imblearn.under_sampling import RandomUnderSampler
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

/usr/local/lib/python3.7/dist-packages/sklearn/externals/six.py:31: FutureWarning: The module is deprecated in version 0.21 and will be removed in version 0.23 since we've dropped support for Python 2.7. Please rely on the official version of six (https://pypi.org/project/six/).
  "(https://pypi.org/project/six/).", FutureWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:144: FutureWarning: The sklearn.neighbors.base module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.neighbors. Anything that cannot be imported from sklearn.neighbors is now part of the private API.
  warnings.warn(message, FutureWarning)


# Step 2: Read Dataset

In [4]:
# Load Dataset

data = pd.read_excel('drive/My Drive/Colab Notebooks/WisconsinDataset/Forms_cleaned.xlsx', 'form1')

print('\n\nDataset:')
print('========\n')
data



Dataset:



EID FormID  Flagged  Attempt  ...  idur.167  idur.168  idur.169  idur.170
0     e100001  form1        0        5  ...        53        82       100        21
1     e100002  form1        1        1  ...       120        93        87        29
2     e100003  form1        1        1  ...        63        96        79        48
3     e100004  form1        1        3  ...        59        41        84        17
4     e100005  form1        1        5  ...       164        74        63        17
...       ...    ...      ...      ...  ...       ...       ...       ...       ...
1631  e101632  form1        0        1  ...        56        47        61        19
1632  e101633  form1        0        1  ...        79        59        48        40
1633  e101634  form1        0        1  ...        19        27        43        14
1634  e101635  form1        0        1  ...        18       109        94        14
1635  e101636  form1        0        1  ...        28        51        45        46

[1636 rows x 515 columns]

# Step 3: Understand Datset

In [5]:
# Understand Dataset

print("\n\nAttributes of Dataset:")
print("======================\n")

print("\nName of Dataset: Forms_cleaned")
print("==============================\n")

print("\nNumber of Rows:", len(data))
print("====================\n")

print("\nNumber of Columns:", len(data.columns))
print("======================\n")



Attributes of Dataset:


Name of Dataset: Forms_cleaned


Number of Rows: 1636


Number of Columns: 515



# Step 4: Pre-Process Dataset

In [6]:
# Pre-Processing

data.tail()
data.drop(data.loc[:, 'iresp.1':'iresp.170'], inplace = True, axis = 1) 
data_final = data

In [7]:
# Count Null Values

print(data_final.isnull().sum())
print(data_final.shape)

EID         0
FormID      0
Flagged     0
Attempt     0
tot_time    0
           ..
idur.166    0
idur.167    0
idur.168    0
idur.169    0
idur.170    0
Length: 345, dtype: int64
(1636, 345)


In [8]:
# Drop Null Values

data_final.dropna(inplace=True)
print(data_final.isnull().sum())
print(data_final.shape)

EID         0
FormID      0
Flagged     0
Attempt     0
tot_time    0
           ..
idur.166    0
idur.167    0
idur.168    0
idur.169    0
idur.170    0
Length: 345, dtype: int64
(1636, 345)


In [9]:
# Time

data_final['mean_idur'] = data_final.loc[:, 'idur.1':'idur.170'].mean(axis=1)
data_final['median_idur'] = data_final.loc[:, 'idur.1':'idur.170'].median(axis=1)
data_final['max_idur'] = data_final.loc[:, 'idur.1':'idur.170'].max(axis=1)
data_final['min_idur'] = data_final.loc[:, 'idur.1':'idur.170'].min(axis=1)

# Correct Answers

data_final['iraw_score'] = data_final.loc[:, 'iraw.1':'iraw.170'].sum(axis=1)
data_final['iraw_score_pct'] = data_final['iraw_score']/170
data_final

EID FormID  Flagged  ...  min_idur  iraw_score  iraw_score_pct
0     e100001  form1        0  ...        14          54        0.317647
1     e100002  form1        1  ...        11          55        0.323529
2     e100003  form1        1  ...        16          61        0.358824
3     e100004  form1        1  ...        12          61        0.358824
4     e100005  form1        1  ...         0          62        0.364706
...       ...    ...      ...  ...       ...         ...             ...
1631  e101632  form1        0  ...         8         156        0.917647
1632  e101633  form1        0  ...         7         153        0.900000
1633  e101634  form1        0  ...         7         154        0.905882
1634  e101635  form1        0  ...        10         152        0.894118
1635  e101636  form1        0  ...        10         156        0.917647

[1636 rows x 351 columns]

# Step 5: Feature Selection

In [10]:
# iraw_idur_OEF (both+OEF)

col_features = ['Attempt', 'tot_time', 'mean_idur', 'median_idur', 'max_idur', 'min_idur', 'iraw_score_pct']

for x in range(1,171):
    col_features.append('idur.'+str(x))

for i in range(1,171):
    col_features.append('iraw.'+str(i))

X = data_final.loc[:, col_features]
Y = data_final.Flagged    

In [11]:
# Count Values

counter = Counter(Y)
print(counter)

Counter({0: 1590, 1: 46})


# Step 6: Feature Scaling

In [12]:
# Scaling

scaler = MinMaxScaler(feature_range=(0, 1))
X_scaled = scaler.fit_transform(X)

In [13]:
# SMOTE and RandomUnderSampler (for balancing data)

over = SMOTE(sampling_strategy=0.4)
under = RandomUnderSampler(sampling_strategy=0.5)
steps = [('o', over), ('u', under)]

pipeline = Pipeline(steps=steps)

<center>
    <br>
        <br>
            <br>
                <h5 style = "color:blue">====================================================================</h5>
                <h2 style = "color:green">Hyper Parameter Tuning</h2>
                <h5 style = "color:green">Training Classifiers using Hyperparameter Optimization as Genetic Algorithms</h5>
                <h5 style = "color:blue">====================================================================</h5>
            <br>
        <br>
    <br>
</center>

**(before stacking)** <br><br>
Tuning seperately: **Random Forest**, Gradient Boosting, Naive Bayes, SVM 



In [14]:
# Hyper Parameter Tunning  --- Random Forest

x_all, y_all = pipeline.fit_resample(X_scaled, Y)

# Number of trees in random forest
n_estimators = [int(x) for x in np.linspace(start = 200, stop = 1000, num = 10)]

# Number of features to consider at every split
max_features = ['auto', 'sqrt','log2']

# Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(10, 500,10)]

# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10]

# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 4,6]

# Create the random grid
param1 = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
              'criterion':['entropy','gini']}
print(param1)

clf1 = TPOTClassifier(generations= 5, population_size= 24, offspring_size= 12,
                                 verbosity= 2, early_stop= 12,
                                 config_dict={'sklearn.ensemble.RandomForestClassifier': param1}, 
                                 cv = 4, scoring = 'accuracy')

clf1.fit(x_all, y_all)
clf1.export('best_rf.py')

{'n_estimators': [200, 288, 377, 466, 555, 644, 733, 822, 911, 1000], 'max_features': ['auto', 'sqrt', 'log2'], 'max_depth': [10, 64, 118, 173, 227, 282, 336, 391, 445, 500], 'min_samples_split': [2, 5, 10], 'min_samples_leaf': [1, 2, 4, 6], 'criterion': ['entropy', 'gini']}


Optimization Progress:   0%|          | 0/84 [00:00<?, ?pipeline/s]


Generation 1 - Current best internal CV score: 0.9931865828092243

Generation 2 - Current best internal CV score: 0.9937106918238994

Generation 3 - Current best internal CV score: 0.9937106918238994

Generation 4 - Current best internal CV score: 0.9942348008385744

Generation 5 - Current best internal CV score: 0.9942348008385744

Best pipeline: RandomForestClassifier(input_matrix, criterion=gini, max_depth=500, max_features=auto, min_samples_leaf=1, min_samples_split=2, n_estimators=911)



=============================================================== <br>
**the best:** <br>
RandomForestClassifier(criterion="gini", max_depth=500, max_features="auto", min_samples_leaf=1, min_samples_split=2, n_estimators=911) <br>
=============================================================== <br>